man/unhide-tcp.8
.TH "UNHIDE-TCP" "8" "January 2020" "Administration commands"
.SH "NAME"
unhide-tcp \(em forensic tool to find hidden TCP/UDP ports
.SH "SYNOPSIS"
.PP
\fBunhide-tcp\fR [\fIOPTIONS\fR]
.SH "DESCRIPTION"
.PP
\fBunhide-tcp\fR is a forensic tool that identifies TCP/UDP ports that are listening but are not listed by /sbin/ss (or alternatively by /bin/netstat) through brute forcing of all TCP/UDP ports avail‐ able.
.PP
Note : If iproute2 is not available on the system, option -n or -s SHOULD be given on the command line
.PP
if unhide process crash you can have a rootkit in the system with some bugs
.PP
.SH "OPTIONS"
.TP
\fB-f, --show-fuser\fR 
show fuser output for hidden ports
.TP
\fB-l, --logfile \fR 
log result into unhide-tcp.log file
.TP
\fB-n, --use-netstat\fR 
use netstat instead of ss
.TP
\fB-o, --show-lsof\fR 
show lsof output for hidden ports
.TP
\fB-s, --use-quickver\fR 
use very quick version for server with lot of opened ports
.TP
\fB-v, --verbose\fR 
verbose
.TP
\fB-?, --help\fR 
Give this help list
.TP
\fB--usage\fR 
Give a short usage message
.TP
\fB-V, --version\fR 
Print program version
.PP
.SS "Exit status:"
.TP
value different from 4, 8, or 12
if no hidden port is found,
.TP
4
if one or more hidden TCP port(s) is(are) found,
.TP
8
if one or more hidden UDP port(s) is(are) found,
.TP
12
if one or more hidden TCP and UDP ports are found.
.PP
.SH "EXAMPLES"
.TP
Standard:
unhide-tcp 
.TP
Other example:
unhide-tcp --show-fuser --show-lsof --logfile
.SH "BUGS"
.PP
Report \fBunhide\fR bugs on https://github.com/YJesus/Unhide
.br
via email dreg@fr33project.org - yjesus@security-projects.com
.SH "SEE ALSO"
.PP
unhide-tcp(8), unhide-gids(8), unhide(8), unhide-posix(8)
.SH "AUTHOR"
First version manual page was written by Francois Marier francois@debian.org and Patrick Gouin. 
.br
The current version was written by David Reguera Garcia aka Dreg.
.br
Permission is granted to copy, distribute and/or modify this document under
the terms of the GNU General Public License, Version 3 or any
later version published by the Free Software Foundation.
.SH LICENSE
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


NEWS
Changes since 20110113 :
**********************

IMPORTANT

  - unhide-linux26.c was renamed to unhide-linux.c
  - unhide.c was renamed to unhide-posix.c
  - The log file of unhide-linux is renamed 'unhide-linux_AAAA-MM-DD.log'
  - The log file of unhide-tcp is named 'unhide-tcp_AAAA-MM-DD.log'
  - By default, unhide-tcp now use /sbin/ss from iproute2 package, to use
    netstat as before '-n' option must be given on command line.
  - Display is more verbose and multi-lines for hidden processes (unhide-linux).
  - If asked to (-l and/or -f), display is more verbose and multi-lines for hidden ports (unhide-tcp).
  - sysinfo test is no more called as part of compound quick and sys tests as it may give false positives.
    It could still be run using the checksysinfo, checksysinfo2 or checksysinfo3 command line parameter.

NEW FEATURES

  - Major enhancement of unhide-tcp :
     * Add capability to output a log file (unhide-tcp_AAA-MM-DD.log)
     * Add capability to output more information (via lsof and/or fuser) on hidden port if available
     * Add verbose mode (disabled by default) to display warning
     * Add a new method (via option '-s') very fast on system with huge number of opened ports
     * Make a double check of port access to avoid false positive (previous single check
       version is available as unhide-tcp-simple-check.c if needed).
  - Add a quick port in C language of unhide.rb (unhide_rb.c) and guess what ...
    it's 40 times faster than original ruby unhide.rb
    unhide_rb doesn't take any option.
  - Add "-d" option for doing a double check in brute test, this reduce false positives.
  - Add "-o" option as synonym of "-f".
  - For found hidden processes, display the user and the working directory
    as extracted from the process environment. Note that it doesn't work well 
    for kernel processes/threads nor for deamons.
  - For found hidden processes, display cmdline, exe link and internal command name.

MISCELLANOUS

  - Add french and spanish man page for unhide-tcp
  - Update english manpage of unhide-tcp to reflect changes
  - Minor corrections in french manpage of unhide
  - Display copyright and license information in start banners.
  - Make message from sysinfo tests more clear.
  - Add a NEWS file :)
  - Update README.txt, LISEZ-MOI.txt and LEEME.txt to clarify difference between
    unhide-posix and unhide-linux.
  - Remove sysinfo test from quick and sys compound tests as it may give false positive.
    sysinfo test still can be used via the checksysinfo[2|3] command line parameters.

BUG FIXES

  - Suppress pedantic compilation warnings (glibc >=2.3, gcc >=4.6).
  - Correct the number of processes displayed for /proc counting in sysinfo test.

Changes since 20100819 :
**********************

NEW FEATURES

  - Add spanish man page
  - Add additional check to checkopendir when -m is specified.
  - Add a option (-f) to create a log file.
  - Add checkopendir test (also called by procfs and procall compound test)
  - Also do opendir() test in reverse and quick tests.
  - Add alternate sysinfo test (via -r option or checksysinfo2 test name)
  - Make the output of hidden process on one line to facilitate parsing
  - Display wchan if there is no cmdline and no exe link (sleeping kernel threads)
  - Add -V version to show version and exit.
  - The -v option can now be given more than once on command line : management of several verbosity level.
  - Now several tests can be simultaneously entered on the command line.
  - Add all elementary tests to the command line test list
  - Add procall compound test command line args.
  - Check for our own spawn ps process in reverse test to avoid false positive.
  - Enhanced fake process detection in reverse test.

BUG FIXES

  - Correct warning message in additional check of checkchdir.
  - Close log file only if it is open.
  - Correct the value returned by unhide
  - Add the misssing new lines in most of the warnings (thanks to gordy for the report).
  - Check the return of fgets in checkallreverse(), check of feof seems not to be
     very reliable for a pipe, we sometime got the last line 2 times (thanks to gordy for the report).
  - Correct an initialized fd use, that gcc don't report when -O2 isn't given on command line

DEVELOPER ISSUES

  - Minor readability when generating program info for display
  - Factorize (f)printf to stdout & log.
  - Add a preliminary testsuite for unhide (sanity.sh)
  - Use printbadpid() in checkallnoprocps() as in other tests.
  - Also check it in checksysinfo & checksysinfo2
  - Simplify and clarify test checksysinfo()
  - Redo args parsing : Manage multiple args on command line and several verbosity levels.
  - Add a tests table to allow new command line parsing.
  - Correct a copy/past "typo", in checkps
  - Minor optimizations of printf & sprintf calls.

MISCELLANOUS

  - Add a NEWS file
  - Add GPL disclaimer to source files
  - Add french LISEZ-MOI.txt file
  - Add reference to new unhide site in version string
  - Add a warning about the generic version of unhide in README.txt (thanks to gordy for the report)
  - Modify man page to add the -V option, correct typos and clarify quick test.
  - Add -O2 option to compiling command line in README.txt
  - Add a TODO file


unhide-tcp.c
/*
          http://www.unhide-forensics.info
*/

/*
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <stdio.h>
#include <argp.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <errno.h>
#include <getopt.h>
#include <string.h>
#include <unistd.h>
#include <time.h>
#include <sys/time.h>
#include <sys/resource.h>

#include "unhide-output.h"
#include "unhide-tcp.h"

const char* argp_program_version = "unhide-tcp 20200120";
const char* argp_program_bug_address = "http://www.unhide-forensics.info";

static char doc[] =
"unhide-tcp options:\
\v-";

static char args_doc[] = " ";

// header
const char header[] =
       "Copyright © 2012-2020 Yago Jesus, Patrick Gouin & David Reguera aka Dreg\n"
       "License GPLv3+ : GNU GPL version 3 or later\n"
       "http://www.unhide-forensics.info\n"
       "NOTE : some rootkits detects unhide checking its name. Just copy the original executable with a random name\n"
       "if unhide process crash you can have a rootkit in the system with some bugs\n\n";

// options
int verbose = 0;
int use_fuser = 0;
int use_lsof = 0;
int use_ss = 1;   // use ss by default
int use_quick = 0;

char checker[10] = "ss" ;

// Temporary string for output
char scratch[1000];

// For logging to file
int logtofile = 0;
FILE *unlog;

// Global hidden port counter, used only to set the exit code of the program
int hidden_found;


/* thx aramosf@unsec.net for the nice regexp! */

// Linux
char tcpcommand1[]= "netstat -tan | sed -e '/[\\.:][0-9]/!d' -e 's/.*[\\.:]\\([0-9]*\\) .*[\\.:].*/\\1/'" ;
char udpcommand1[]= "netstat -uan | sed -e '/[\\.:][0-9]/!d' -e 's/.*[\\.:]\\([0-9]*\\) .*[\\.:].*/\\1/'" ;

// Alternative commands, needs iproute2
char tcpcommand2[]= "ss -tan sport = :%d | sed -e '/[\\.:][0-9]/!d' -e 's/.*[\\.:]\\([0-9]*\\) .*[\\.:].*/\\1/'" ;
char udpcommand2[]= "ss -uan sport = :%d | sed -e '/[\\.:][0-9]/!d' -e 's/.*[\\.:]\\([0-9]*\\) .*[\\.:].*/\\1/'" ;

// fuser commands
char fuserTCPcommand[]= "fuser -v -n tcp %d 2>&1" ;
char fuserUDPcommand[]= "fuser -v -n udp %d 2>&1" ;

// lsof commands
char lsofTCPcommand[]= "lsof +c 0 -iTCP:%d" ;
char lsofUDPcommand[]= "lsof +c 0 -iUDP:%d" ;

// OpenBSD
// char tcpcommand[]= "netstat -an -p tcp | sed -e '/[\\.:][0-9]/!d' -e 's/.*[\\.:]\\([0-9]*\\) .*[\\.:].*/\\1/'" ;
// char udpcommand[]= "netstat -an -p udp| sed -e '/[\\.:][0-9]/!d' -e 's/.*[\\.:]\\([0-9]*\\) .*[\\.:].*/\\1/'" ;


// Solaris
// char tcpcommand[]= "netstat -an -P tcp | sed -e '/[\\.:][0-9]/!d' -e 's/.*[\\.:]\\([0-9]*\\) .*[\\.:].*/\\1/'" ;
// char udpcommand[]= "netstat -an -P udp| sed -e '/[\\.:][0-9]/!d' -e 's/.*[\\.:]\\([0-9]*\\) .*[\\.:].*/\\1/'" ;



/*
 *  Run a command to get more information about a port. 
 */
static void print_info(const char *prog_name, const char *command_fmt, int port)
{

   char buffer[1000];
   FILE* fp;

   sprintf(buffer, command_fmt, port);
   fp = popen(buffer, "r") ;

   if (NULL == fp)
   {
      warnln(verbose, unlog, "Couldn't run command: %s", buffer) ;
      return ;
   }

   msgln(unlog, 1, "%s reports :", prog_name) ;

   while (NULL != fgets(buffer, 1000, fp))
   {
      msgln(unlog, 1, buffer) ;
   }

   pclose(fp);
}

/* Print a port, optionally querying info about it via lsof or fuser. */
void print_port(enum Proto proto, int port)
{
      msgln(unlog, 0, "\nFound Hidden port that not appears in %s: %i", checker, port) ;
      if (1 == use_fuser)
      {
         if (TCP == proto)
         {
            print_info("fuser", fuserTCPcommand, port);
         }
         else
         {
            print_info("fuser", fuserUDPcommand, port);
         }
      }
      if (1 == use_lsof)
      {
         if (TCP == proto)
         {
            print_info("lsof", lsofTCPcommand, port);
         }
         else
         {
            print_info("lsof", lsofUDPcommand, port);
         }
      }
}


/*
 * Check if port is seen by netstat.
 *
 * If not, report it and optionnally run lsof and/or fuser
 * to show more info.
 */
int checkoneport(int port, char command[], enum Proto proto)
{
   int ok = 0;
   char ports[30];
   char compare[100];

   FILE *fich_tmp ;

   if (NULL != (fich_tmp=popen (command, "r")))
   {
      sprintf(compare,"%i\n",port);
      while ((NULL != fgets(ports, 30, fich_tmp)) && ok == 0) {
         if (strcmp(ports, compare) == 0) {ok = 1;}
      }
      pclose(fich_tmp);
   }
   else
   {
      die(unlog, "Couldn't execute command : %s while checking port %d", command, port) ;
   }
   return(ok) ;
}

/*
 * Check all TCP ports one by one.
 */
static void print_hidden_TCP_ports_1_by_1(enum Proto proto)
{
   int i ;
   char tcpcommand[512] ;

   hidden_found = 0 ;
   for (i =1; i <= 65535; i++)
   {
      int socket_desc;
      struct sockaddr_in address;

      if ( -1 != (socket_desc=socket(AF_INET,SOCK_STREAM,0)))
      {
         address.sin_family = AF_INET;
         address.sin_addr.s_addr = INADDR_ANY;
         address.sin_port = htons(i);
         errno= 0 ;
         if ( -1 != bind(socket_desc,(struct sockaddr *)&address,sizeof(address)))
         {
            listen(socket_desc,1);
            if ( EADDRINUSE == errno )    // port is listened by another process
            {
               if (1 == use_ss)
               {
                  sprintf(tcpcommand, tcpcommand2, i) ;
               }
               else
               {
                  strncpy(tcpcommand, tcpcommand1, 512) ;
               }
               if (0 == checkoneport(i, tcpcommand, TCP))
               {
                  // test again
                  listen(socket_desc,1);
                  if ( EADDRINUSE == errno )    // port is still listened by another process
                  {
                     hidden_found++;
                     print_port(proto, i) ;
                  }
               }
               close(socket_desc);
            }
            else
            {
               close(socket_desc);
            }
         }
         else
         {
            if (EADDRINUSE == errno)    //port is in use by another process
            {
               if (1 == use_ss)
               {
                  sprintf(tcpcommand, tcpcommand2, i) ;
               }
               else
               {
                  strncpy(tcpcommand, tcpcommand1, 512) ;
               }
               if (0 == checkoneport(i, tcpcommand, TCP))
               {
                  // test again
                  if ( -1 == bind(socket_desc,(struct sockaddr *)&address,sizeof(address)))
                  {
                     if ( EADDRINUSE == errno )    // port is still used by another process
                     {
                        hidden_found++;
                        print_port(proto, i) ;
                     }
                     else
                     {
                        warnln(verbose, unlog, "can't bind to socket while checking port %d", i) ;
                     }
                     close(socket_desc);
                  }
               }
               else
               {
                  close(socket_desc);
               }
            }
         }
      }
      else
      {
         warnln(verbose, unlog, "can't create socket while checking port %d/tcp", i) ;
      }
   }
}

/*
 * Check all UDP ports one by one.
 */
static void print_hidden_UDP_ports_1_by_1(enum Proto proto)
{
   int u ;
   char udpcommand[512] ;

   hidden_found = 0 ;

   for (u = 1; u <= 65535; u++)
   {
      int socket_desc;
      struct sockaddr_in address;

      if ( -1 != (socket_desc=socket(AF_INET,SOCK_DGRAM,0)))
      {
         address.sin_family = AF_INET;
         address.sin_addr.s_addr = INADDR_ANY;
         address.sin_port = htons(u);
         errno= 0 ;
         if ( 0 != bind(socket_desc,(struct sockaddr *)&address,sizeof(address)))
         {
            if ( EADDRINUSE == errno )   //port is in use by another process
            {
               if (1 == use_ss)
               {
                  sprintf(udpcommand, udpcommand2, u) ;
               }
               else
               {
                  strncpy(udpcommand, udpcommand1, 512) ;
               }
               ;
               if (0 == checkoneport(u, udpcommand, UDP))
               {
                  // test again
                  if ( 0 != bind(socket_desc,(struct sockaddr *)&address,sizeof(address)))    // port is still in use by another process
                  {
                     if ( EADDRINUSE == errno )   //port is in use by another process
                     {
                        hidden_found++;
                        print_port(proto, u) ;
                     }
                  }
               }
               close(socket_desc);
            }
            else      // other error
            {
               close(socket_desc);
               warnln(verbose, unlog, "can't bind to socket while checking port %d", u) ;
            }
         }
         else  // port is available
         {
            close(socket_desc);
         }
      }
      else 
      {
         warnln(verbose, unlog, "can't create socket while checking port %d/udp", u) ;
      }
   }
}

static struct argp_option options[] =
{
	{ "show-fuser", 'f', 0, OPTION_ARG_OPTIONAL, "show fuser output for hidden ports" },
	{ "show-lsof", 'o', 0, OPTION_ARG_OPTIONAL, "show lsof output for hidden ports" },
	{ "use-quickver", 's', 0, OPTION_ARG_OPTIONAL, "use very quick version for server with lot of opened ports" },
	{ "use-netstat", 'n', 0, OPTION_ARG_OPTIONAL, "use netstat instead of ss" },
	{ "logfile", 'l', 0, OPTION_ARG_OPTIONAL, "log result into unhide-tcp.log file", 0 },
	{ "verbose", 'v', 0, OPTION_ARG_OPTIONAL, "verbose", 0 },
	{ NULL, 0, NULL, 0, NULL, 0 }
};

static error_t parse_opt(int key, char* arg, struct argp_state* state)
{
	struct arguments* arguments = (struct arguments*) state->input;

	switch (key)
	{
	case 'v':
		verbose = 1;
		break;
		
	case 'l': 
		logtofile = 1;
		break;
		
	case 'f': 
		use_fuser = 1 ;
		break;
		
	case 'o': 
		use_lsof = 1 ;
		break;
		
	case 's': 
		use_quick = 1 ;
		break;
		
	case 'n': 
		use_ss = 0 ;
		break;

	case ARGP_KEY_END:
		break;

	default:
		return ARGP_ERR_UNKNOWN;
	}

	return 0;
}

static struct argp argp = { options, parse_opt, args_doc, doc, NULL, NULL, NULL };


/*
 * Look for TCP and UDP ports that are hidden to netstat.
 *
 * Returns 0 if none is found, 1 if there is some internal error, 4 if TCP
 * hidden ports were found, 8 if UDP hidden ports were found or 12 (4 & 8) if
 * both were found.
 */
int main(int argc, char  **argv) 
{
   int ret_code = 0;

   printf("%s\n%s", argp_program_version, header);

   if(getuid() != 0){
      die(unlog, "You must be root to run %s !", argv[0]) ;
   }
   
   argp_parse(&argp, argc, argv, 0, 0, NULL);   
   
   if (1 == logtofile) 
   {
      unlog = init_log(logtofile, header, "unhide-tcp") ;
   }

   if (1 == use_ss)
   {
      strncpy(checker, "ss", 10);
   }
   else
   {
      strncpy(checker, "netstat", 10);
   }
   
   setpriority(PRIO_PROCESS,0,-20);  /* reduce risk of race condition - may fail, dont care */
   
   msgln(unlog, 0, "[*]Starting TCP checking") ;
   if (1 == use_quick)
   {
      print_hidden_ports(TCP);
   }
   else
   {
      print_hidden_TCP_ports_1_by_1(TCP) ;
   }
   if (hidden_found)
   {
      ret_code += 4;
   }

   msgln(unlog, 0, "[*]Starting UDP checking") ;
   if (1 == use_quick)
   {
      print_hidden_ports(UDP);
   }
   else
   {
      print_hidden_UDP_ports_1_by_1(UDP) ;
   }
   if (hidden_found)
   {
      ret_code += 8;
   }

   if (logtofile == 1)
   {
      close_log(unlog, "unhide-tcp") ;
   }
   return(ret_code);

}

